README.md
# NAME

CPAN::Repository - API to access a directory which can be served as CPAN repository

[![Build Status](https://travis-ci.org/Getty/p5-cpan-repository.png?branch=master)](https://travis-ci.org/Getty/p5-cpan-repository)

# VERSION

version 0.011

# SYNOPSIS

     use CPAN::Repository;

     my $repo = CPAN::Repository->new({
       dir => '/var/www/greypan.company.org/htdocs/',
       url => 'http://greypan.company.org/',
     });
     
     $repo->initialize unless $repo->is_initialized;
     
     $repo->add_author_distribution('AUTHOR','My-Distribution-0.001.tar.gz');
     $repo->add_author_distribution('AUTHOR2','Other-Dist-0.001.tar.gz','Custom/Own/Path');
     $repo->set_alias('AUTHOR','The author <author@company.org>');
     $repo->set_alias('AUTHOR2','The other author <author@company.org>');
     
     my %modules = %{$repo->modules};
     
     my $fullpath_to_authordir = $repo->authordir('SOMEONE');
    
     my $packages = $repo->packages; # gives back a CPAN::Repository::Packages
     my $mailrc = $repo->mailrc; # gives back a CPAN::Repository::Mailrc

# DESCRIPTION

This module is made for representing a directory which can be used as own CPAN for modules, so it can be a GreyPAN, a DarkPAN or even can be
used to manage a mirror of real CPAN your own way. Some code parts are taken from CPAN::Dark of **CHROMATIC** and [CPAN::Mini::Inject](https://metacpan.org/pod/CPAN::Mini::Inject) of **MITHALDU**.

# SEE ALSO

[CPAN::Repository::Packages](https://metacpan.org/pod/CPAN::Repository::Packages)

[CPAN::Repository::Mailrc](https://metacpan.org/pod/CPAN::Repository::Mailrc)

# SUPPORT

IRC

    Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

    http://github.com/Getty/p5-cpan-repository
    Pull request and additional contributors are welcome

Issue Tracker

    http://github.com/Getty/p5-cpan-repository/issues

# AUTHOR

Torsten Raudssus <torsten@raudss.us> [http://raudss.us/](http://raudss.us/)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by DuckDuckGo, Inc. [http://duckduckgo.com/](http://duckduckgo.com/).

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
==================================================
Changes from 1742-02-20 00:00:00 +0000 to present.
==================================================

{{$NEXT}}


------------------------------------------
version 0.010 at 2015-12-05 04:05:28 +0000
------------------------------------------

  Change: 98a84303921abb9841d179c88499cfa826ad8a68
  Author: Olaf Alders <olaf@wundersolutions.com>
  Date : 2015-12-04 23:03:20 +0000

    Ensure CPAN::Repository always has a $VERSION 

------------------------------------------
version 0.009 at 2015-04-23 00:56:50 +0000
------------------------------------------

  Change: c7a1864d46c47229c975ff210eabd251343addc4
  Author: Olaf Alders <olaf@wundersolutions.com>
  Date : 2015-04-22 20:56:50 +0000

    Commit dzil's changes to Travis config. 

  Change: 0fa21891269de0e609692d0ce3635a09f5f9b116
  Author: Olaf Alders <olaf@wundersolutions.com>
  Date : 2015-04-22 20:56:50 +0000

    Ensure that only one permission is written for each line. 

  Change: 985ebce92a2bf33b6d8507ff3fab32c3f5465dec
  Author: Olaf Alders <olaf@wundersolutions.com>
  Date : 2015-04-22 20:55:19 +0000

    Merge pull request #1 from ivanwills/master

    Fixed "self" refernces to be "repo" to match reset of the
    documentation 

  Change: 8a62b1a824847d5c302b967e801c06f4f2e7c096
  Author: Ivan Wills <ivan.wills@gmail.com>
  Date : 2014-12-25 18:55:31 +0000

    Fixed "self" refernces to be "repo" to match reset of the
    documentation 

  Change: 58a2180ad00e4ec4a2e31c984fdc0ea2e7b045d8
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-05-02 23:24:14 +0000

    New .travis.yml 

  Change: fda31b41a17077c982067030d278dcfe8db1e1f6
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-04-25 01:22:22 +0000

    Committed updated travis.yml 

  Change: 8816481cf18293611f73bd3e0c3752b8402a2c96
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-04-25 01:21:22 +0000

    Added travis.yml 

------------------------------------------
version 0.008 at 2013-02-19 04:09:09 +0000
------------------------------------------

  Change: 751bb68a9cc2167f37f9e7ba3e122732cbcfec34
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2013-02-19 05:09:09 +0000

    Added possibility for custom author path, tuned tests a bit 

  Change: 826b2b1016705a4dd35664ed2d38e1378d4b3377
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:22:08 +0000

    A bit more handles on CPAN::Repository, a bit more informational
    functions 

  Change: d0fd8a4a28dfc85f2c5d00d7e07484a49bc49ff9
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:09:21 +0000

    Hups, missed generation a bit ;) 

  Change: 817fc4ca199e2ed86904a541aea4d3f25f1fdc42
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 17:06:57 +0000

    Added 06perms.txt 

  Change: 638fbb234a31dbad87c44dd518b05edbffedf9de
  Author: Torsten Raudssus <torsten@raudss.us>
  Date : 2012-07-09 16:29:41 +0000

    Added sorted output to 02packages file 

------------------------------------------
version 0.007 at 2012-01-26 00:33:00 +0000
------------------------------------------

  Change: 40882f1e46e8a07633387632621ef6958b9989d6
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:33:00 +0000

    v0.007 

  Change: 07f6d26a55d513de7cc6a3fa0201086e3885905a
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:31:50 +0000

    Damn it, remove debugging before release! 

------------------------------------------
version 0.006 at 2012-01-26 00:23:36 +0000
------------------------------------------

  Change: fc73dd923155974b0ecbaaa1a392805de467dbf1
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:23:36 +0000

    v0.006

    Fixed handling of undef versions in generation of packages 

  Change: 4a55e90e39dee509b9dfc520f187b77a575fa46e
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-26 01:07:32 +0000

    Fixed handling of undef versions in generation of packages 

------------------------------------------
version 0.005 at 2012-01-19 17:27:21 +0000
------------------------------------------

  Change: 6805f253673872c2c33d7e7a61049401b5cca191
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-19 18:27:21 +0000

    v0.005

    Return pathname in repo on add_author_distribution 

  Change: 686c92891b684192140ed705ceb8748600bf87b6
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-19 18:25:58 +0000

    Return pathname in repo on add_author_distribution 

------------------------------------------
version 0.004 at 2012-01-18 00:01:32 +0000
------------------------------------------

  Change: 282f22ce49f2842a23a144cef129d58d16aa463c
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-18 01:01:32 +0000

    v0.004

    Fixed documentation, added generation of 02STAMP 

  Change: 2fcd4dea3741769ccd95fe5c423340bc1d0566f1
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-18 01:00:34 +0000

    Added timestamp feature, added more pod (thats always good) 

  Change: 222fb80936d073994ae6c0cba3a9a1b0ed509183
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:17:02 +0000

    Fixed typo in SYNOPSIS and a stupid 1; at the end of pod 

------------------------------------------
version 0.003 at 2012-01-17 03:12:15 +0000
------------------------------------------

  Change: 20f044868e11d112bc06a9e6112e239f3b438831
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:12:15 +0000

    v0.003

    Arghl... another requirement fix 

  Change: 31a77af4edb34f830ec8be8c68b637f186a56bfa
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:11:09 +0000

    And again forgot Changes... arghl^2 

  Change: f6dd17fb84ba3d3b2531e46dbbf73f56513da665
  Author: Torsten Raudssus <github@raudssus.de>
  Date : 2012-01-17 04:10:49 +0000

    Arghl.. another requirement fix... 

------------------------------------------

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "API to access a directory which can be served as CPAN repository",
  "AUTHOR" => "Torsten Raudssus <torsten\@raudss.us> L<http://raudss.us/>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CPAN-Repository",
  "LICENSE" => "perl",
  "NAME" => "CPAN::Repository",
  "PREREQ_PM" => {
    "DateTime" => "0.72",
    "DateTime::Format::Epoch" => "0.13",
    "DateTime::Format::Epoch::Unix" => 0,
    "DateTime::Format::RFC3339" => 0,
    "Dist::Data" => "0.002",
    "File::Copy" => 0,
    "File::Path" => "2.08",
    "File::Spec::Functions" => "3.33",
    "IO::File" => "1.14",
    "IO::Zlib" => "1.10",
    "Moo" => "0.009013",
    "Moo::Role" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => "0.22",
    "FindBin" => 0,
    "Test::More" => "0.90",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DateTime" => "0.72",
  "DateTime::Format::Epoch" => "0.13",
  "DateTime::Format::Epoch::Unix" => 0,
  "DateTime::Format::RFC3339" => 0,
  "Dist::Data" => "0.002",
  "ExtUtils::MakeMaker" => 0,
  "File::Copy" => 0,
  "File::Path" => "2.08",
  "File::Spec" => 0,
  "File::Spec::Functions" => "3.33",
  "File::Temp" => "0.22",
  "FindBin" => 0,
  "IO::File" => "1.14",
  "IO::Zlib" => "1.10",
  "Moo" => "0.009013",
  "Moo::Role" => 0,
  "Test::More" => "0.90",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/CPAN/Repository/Perms.pm
package CPAN::Repository::Perms;
# ABSTRACT: 06perms

our $VERSION = '0.011';

use Moo;

with qw(
	CPAN::Repository::Role::File
);

use Dist::Data;
use File::Spec::Functions ':ALL';
use IO::File;
use DateTime::Format::RFC3339;
use DateTime::Format::Epoch::Unix;

sub file_parts { 'modules', '06perms.txt' }

our @perm_perms = qw( m f c );

has perms => (
	is => 'rw',
	lazy => 1,
	builder => '_build_perms',
);

sub _build_perms {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %perms;
	for (@lines) {
		chomp($_);
		next if ($_ =~ /[^:]:[ \t]/);
		next if ($_ =~ /^ +/);
		next unless $_;
		my @perm_parts = split(',',$_);
		if (@perm_parts == 3) {
			my $module = shift @perm_parts;
			my $userid = shift @perm_parts;
			my $perm = shift @perm_parts;
			$perms{$module} = {} unless defined $perms{$module};
			warn "already found entry for ".$userid." on ".$module if defined $perms{$module}->{$userid};
			$perms{$module}->{$userid} = $perm;
		}
	}
	return \%perms;
}

has written_by => (
	is => 'ro',
	required => 1,
);

sub perms_linecount {
	my ( $self ) = @_;
	my $i = 0;
	for (keys %{$self->perms}) {
		for my $u (keys %{$self->perms->{$_}}) {
			$i++;
		}
	}
	return $i;
}

sub set_perms {
	my ( $self, $module, $userid, $perm ) = @_;
	die "unknown perm ".$perm unless grep { $_ eq $perm } @perm_perms;
	$self->perms->{$module} = {} unless defined $self->perms->{$module};
	$self->perms->{$module}->{$userid} = $perm;
	return $self;
}

sub get_perms {
	my ( $self, $module ) = @_;
	return defined $self->perms->{$module}
		? $self->perms->{$module}
		: {}
}

sub get_perms_by_userid {
	my ( $self, $userid ) = @_;
	my %perms_by_userid;
	for (keys %{$self->perms}) {
		if (defined $self->perms->{$_}->{$userid}) {
			$perms_by_userid{$_} = $self->perms->{$_}->{$userid};
		}
	}
	return \%perms_by_userid;
}

# File:        06perms.txt
# Description: CSV file of upload permission to the CPAN per namespace
#     best-permission is one of "m" for "modulelist", "f" for
#     "first-come", "c" for "co-maint"
# Columns:     package,userid,best-permission
# Line-Count:  215301
# Written-By:  PAUSE version 1.14
# Date:        Fri, 06 Jul 2012 20:23:21 GMT

sub generate_content {
	my ( $self ) = @_;
	my @file_parts = $self->file_parts;
	my $content = "";
	$content .= $self->generate_header_line('File:',(pop @file_parts));
	$content .= $self->generate_header_line('Description:','Description: CSV file of upload permission to the CPAN per namespace');
	$content .= '    best-permission is one of "m" for "modulelist", "f" for'."\n";
	$content .= '    "first-come", "c" for "co-maint"'."\n";
	$content .= $self->generate_header_line('Columns:','package, userid, best-permission');
	$content .= $self->generate_header_line('Intended-For:','Automated fetch routines, namespace documentation.');
	$content .= $self->generate_header_line('Written-By:',$self->written_by);
	$content .= $self->generate_header_line('Line-Count:',$self->perms_linecount);
	$content .= $self->generate_header_line('Date:',DateTime->now->strftime('%a, %e %b %y %T %Z'));
	$content .= "\n";

    my @perms;
	for (sort { $a cmp $b } keys %{$self->perms}) {
		for my $u (sort { $a cmp $b } keys %{$self->perms->{$_}}) {
			push @perms, join(",",$_,$u,$self->perms->{$_}->{$u});
		}
	}
	$content .= join "\n", @perms;
	return $content;
}

sub generate_header_line {
	my ( $self, $key, $value ) = @_;
	return sprintf("%-13s %s\n",$key,$value);
}

1;

=encoding utf8

=head1 SYNOPSIS

  use CPAN::Repository::Perms;

  my $packages = CPAN::Repository::Perms->new({
    repository_root => $fullpath_to_root,
    written_by => $written_by,
  });

=head1 SEE ALSO
  
L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome
 
Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues


lib/CPAN/Repository/Mailrc.pm
package CPAN::Repository::Mailrc;
# ABSTRACT: 01mailrc

use Moo;

our $VERSION = '0.011';

with qw(
	CPAN::Repository::Role::File
);

sub file_parts { 'authors', '01mailrc.txt' }

has aliases => (
	is => 'ro',
	lazy => 1,
	builder => '_build_aliases',
);

sub _build_aliases {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %aliases;
	for (@lines) {
		if ($_ =~ m/^alias (\w+) "(.*)"$/) {
			$aliases{$1} = $2;
		}
	}
	return \%aliases;
}

sub set_alias {
	my ( $self, $author, $alias ) = @_;
	$self->aliases->{$author} = $alias;
	return $self;
}

sub get_alias {
	my ( $self, $author ) = @_;
	return defined $self->aliases->{$author}
		? $self->aliases->{$author}
		: ();
}

sub generate_content {
	my ( $self ) = @_;
	my $content = "";
	for (sort { $a cmp $b } keys %{$self->aliases}) {
		$content .= 'alias '.$_.' "'.( $self->aliases->{$_} ? $self->aliases->{$_} : $_ ).'"'."\n";
	}
	return $content;
}

1;


=encoding utf8

=head1 SYNOPSIS

  use CPAN::Repository::Mailrc;

  my $mailrc = CPAN::Repository::Mailrc->new({
    repository_root => $fullpath_to_root,
  });

=head1 SEE ALSO
  
L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome
 
Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues


lib/CPAN/Repository/Packages.pm
package CPAN::Repository::Packages;
# ABSTRACT: 02packages

use Moo;

our $VERSION = '0.011';

with qw(
	CPAN::Repository::Role::File
);

use Dist::Data;
use File::Spec::Functions ':ALL';
use IO::File;
use DateTime::Format::RFC3339;
use DateTime::Format::Epoch::Unix;

sub file_parts { 'modules', '02packages.details.txt' }
sub file_parts_stamp { 'modules', '02STAMP' }

has modules => (
	is => 'ro',
	lazy => 1,
	builder => '_build_modules',
);

sub _build_modules {
	my ( $self ) = @_;
	return {} unless $self->exist;
	my @lines = $self->get_file_lines;
	my %modules;
	for (@lines) {
		chomp($_);
		next if ($_ =~ /[^:]:[ \t]/);
		if ($_ =~ m/^([^ \t]+)[ \t]+([^ \t]+)[ \t]+([^ \t]+)$/) {
			# $1 = module
			# $2 = version
			# $3 = path (inside repository)
			$modules{$1} = [ $2, $3 ];
		}
	}
	return \%modules;
}

has authorbase_path_parts => (
	is => 'ro',
	required => 1,
);

has url => (
	is => 'ro',
	required => 1,
);

has written_by => (
	is => 'ro',
	required => 1,
);

sub get_module {
	my ( $self, $module ) = @_;
	return defined $self->modules->{$module}
		? $self->modules->{$module}->[1]
		: ();
}

sub get_module_version {
	my ( $self, $module ) = @_;
	return defined $self->modules->{$module}
		? $self->modules->{$module}->[0]
		: ();
}

sub set_module {
	my ( $self, $module, $version, $path ) = @_;
	return $self->modules->{$module} = [ $version, $path ];
}

sub add_distribution {
	my ( $self, $author_distribution_path ) = @_;
	my $filename = catfile( $self->repository_root, @{$self->authorbase_path_parts}, splitdir( $author_distribution_path ) );
	my $dist = Dist::Data->new( $filename );
	for (keys %{$dist->packages}) {
		$self->set_module($_, $dist->packages->{$_}->{version}, File::Spec::Unix->catfile(splitdir $author_distribution_path));
	}
	return $self;
}

sub stamp_filename {
	my ( $self ) = @_;
	catfile( $self->repository_root, $self->file_parts_stamp );
}

after save => sub {
	my ( $self ) = @_;
	my $stamp = IO::File->new($self->stamp_filename, "w") or die "cant write to ".$self->stamp_filename;
	my $now = DateTime->now;
	print $stamp (DateTime::Format::Epoch::Unix->format_datetime($now).' '.DateTime::Format::RFC3339->new->format_datetime($now)."\n");
	$stamp->close;
};

sub timestamp {
	my ( $self ) = @_;
	my $stamp = IO::File->new($self->stamp_filename, "r") or die "cant read ".$self->stamp_filename;
	my ( $line ) = <$stamp>;
	chomp($line);
	if ($line =~ /^(\d+) /) {
		return DateTime::Format::Epoch::Unix->parse_datetime($1);
	} else {
		die "cant find unix timestamp from ".$self->stamp_filename;
	}
}

sub generate_content {
	my ( $self ) = @_;
	my @file_parts = $self->file_parts;
	my $content = "";
	$content .= $self->generate_header_line('File:',(pop @file_parts));
	$content .= $self->generate_header_line('URL:',$self->url.$self->path_inside_root);
	$content .= $self->generate_header_line('Description:','Package names found in directory $CPAN/authors/id/');
	$content .= $self->generate_header_line('Columns:','package name, version, path');
	$content .= $self->generate_header_line('Intended-For:','Automated fetch routines, namespace documentation.');
	$content .= $self->generate_header_line('Written-By:',$self->written_by);
	$content .= $self->generate_header_line('Line-Count:',scalar keys %{$self->modules});
	$content .= $self->generate_header_line('Last-Updated:',DateTime->now->strftime('%a, %e %b %y %T %Z'));
	$content .= "\n";
	for (sort { $a cmp $b } keys %{$self->modules}) {
		$content .= sprintf("%-60s %-20s %s\n",$_,$self->modules->{$_}->[0] ? $self->modules->{$_}->[0] : 'undef',$self->modules->{$_}->[1]);
	}
	return $content;
}

sub generate_header_line {
	my ( $self, $key, $value ) = @_;
	return sprintf("%-13s %s\n",$key,$value);
}

1;

=encoding utf8

=head1 SYNOPSIS

  use CPAN::Repository::Packages;

  my $packages = CPAN::Repository::Packages->new({
    repository_root => $fullpath_to_root,
    url => $url,
    written_by => $written_by,
    authorbase_path_parts => ['authors','id'],
  });

=head1 SEE ALSO
  
L<CPAN::Repository>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome
 
Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues


lib/CPAN/Repository/Role/File.pm
package CPAN::Repository::Role::File;
# ABSTRACT: Role for file functions

use Moo::Role;

our $VERSION = '0.011';

use File::Path qw( make_path );
use File::Spec::Functions ':ALL';
use IO::Zlib;
use IO::File;

requires qw( file_parts generate_content );

has repository_root => (
	is => 'ro',
	required => 1,
);

has generate_uncompressed => (
	is => 'ro',
	lazy => 1,
	builder => '_build_generate_uncompressed',
);

sub _build_generate_uncompressed { 1 }

sub path_inside_root {
	my ( $self ) = @_;
	return join("/",$self->file_parts);
}

sub compressed_path_inside_root {
	my ( $self ) = @_;
	return join("/",$self->file_parts).".gz";
}

sub full_filename {
	my ( $self ) = @_;
	return catfile( splitdir($self->repository_root), $self->file_parts );
}

sub full_compressed_filename { shift->full_filename.".gz" }

sub exist {
	my ( $self ) = @_;
	return 0 unless -f $self->full_compressed_filename;
	return 1;
}

sub save {
	my ( $self ) = @_;
	my @pps = $self->file_parts;
	pop(@pps);
	$self->mkdir( splitdir( $self->repository_root ), @pps ) unless -d catdir( $self->repository_root, @pps );
	my $content = $self->generate_content;
	my $gz = IO::Zlib->new($self->full_compressed_filename, "w") or die "cant write to ".$self->full_compressed_filename;
	print $gz $content;
	$gz->close;
	if ($self->generate_uncompressed) {
		my $txt = IO::File->new($self->full_filename, "w") or die "cant write to ".$self->full_filename;
		print $txt $content;
		$txt->close;
	}
	return 1;
}

sub get_file_lines {
	my ( $self ) = @_;
	my $gz = IO::Zlib->new($self->full_compressed_filename, "r") or die "cant read ".$self->full_compressed_filename;
	return <$gz>;
}

sub mkdir {
	my ( $self, @path ) = @_;
	make_path(catdir(@path),{ error => \my $err });
	if (@$err) {
		for my $diag (@$err) {
			my ($file, $message) = %$diag;
			if ($file eq '') {
				die "general error: $message\n";
			} else {
				die "problem making path $file: $message\n";
			}
		}
	}
}

1;

=encoding utf8

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome
 
Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues


lib/CPAN/Repository.pm
package CPAN::Repository;
# ABSTRACT: API to access a directory which can be served as CPAN repository

use Moo;

our $VERSION = '0.011';

use File::Path qw( make_path );
use File::Spec::Functions ':ALL';
use CPAN::Repository::Mailrc;
use CPAN::Repository::Packages;
use CPAN::Repository::Perms;
use File::Copy;

has dir => (
	is => 'ro',
	required => 1,
);

has real_dir => (
	is => 'ro',
	lazy => 1,
	builder => '_build_real_dir',
);

sub _build_real_dir { catdir(splitdir(shift->dir)) }

sub splitted_dir { splitdir(shift->real_dir) }

has url => (
	is => 'ro',
	lazy => 1,
	builder => '_build_url',
);

sub _build_url { 'http://cpan.perl.org/' }

has written_by => (
	is => 'ro',
	lazy => 1,
	builder => '_build_written_by',
);

sub _build_written_by { (ref shift).' '.$VERSION }

has mailrc => (
	is => 'ro',
	lazy => 1,
	builder => '_build_mailrc',
	handles => [qw(
		get_alias
	)],
);

sub _build_mailrc {
	my ( $self ) = @_;
	return CPAN::Repository::Mailrc->new({
		repository_root => $self->real_dir,
	});
}

has perms => (
	is => 'ro',
	lazy => 1,
	builder => '_build_perms',
	handles => [qw(
		get_perms
		get_perms_by_userid
	)],
);

sub _build_perms {
	my ( $self ) = @_;
	return CPAN::Repository::Perms->new({
		repository_root => $self->real_dir,
		written_by => $self->written_by,
	});
}


has packages => (
	is => 'ro',
	lazy => 1,
	builder => '_build_packages',
	handles => [qw(
		get_module
		get_module_version
	)],
);

sub _build_packages {
	my ( $self ) = @_;
	return CPAN::Repository::Packages->new({
		repository_root => $self->real_dir,
		url => $self->url,
		written_by => $self->written_by,
		authorbase_path_parts => [$self->authorbase_path_parts],
	});
}

sub is_initialized {
	my ( $self ) = @_;
	$self->mailrc->exist && $self->packages->exist;
}

sub initialize {
	my ( $self ) = @_;
	die "there exist already a repository at ".$self->real_dir if $self->is_initialized;
	$self->mailrc->save;
	$self->packages->save;
	$self->perms->save;
}

sub add_author_distribution {
	my ( $self, $author, $distribution_filename, $path ) = @_;

        die "missing author pause id" unless defined $author;
        die "missing distribution file name" unless defined $distribution_filename;

	my @fileparts = splitdir( $distribution_filename );
	my $filename = pop(@fileparts);
	my $author_path_filename;
	my $target_dir = $self->mkauthordir($author);
	if ($path) {
		my $path_dir = catfile( $self->splitted_dir, $self->authorbase_path_parts, $path );
		$self->mkdir( $path_dir ) unless -d $path_dir;
		$author_path_filename = catfile( $path, $filename );
	} else {
		$author_path_filename = catfile( $self->author_path_parts($author), $filename );
	}
	copy($distribution_filename,catfile( $self->splitted_dir, $self->authorbase_path_parts, $author_path_filename ));
	$self->packages->add_distribution($author_path_filename)->save;
	$self->mailrc->set_alias($author)->save unless defined $self->mailrc->aliases->{$author};
	return catfile( $self->authorbase_path_parts, $self->author_path_parts($author), $filename );
}

sub set_perms {
	my $self = shift;
	$self->perms->set_perms(@_)->save;
}

sub set_alias {
	my ( $self, $author, $alias ) = @_;
	$self->mailrc->set_alias($author,$alias)->save;
}

sub mkauthordir {
	my ( $self, $author ) = @_;
	my $authordir = $self->authordir($author);
	$self->mkdir( $authordir ) unless -d $authordir;
	return $authordir;
}

sub author_path_parts {
	my ( $self, $author ) = @_;
	return substr( $author, 0, 1 ), substr( $author, 0, 2 ), $author;
}

sub authorbase_path_parts { 'authors', 'id' }

sub authordir {
	my ( $self, $author ) = @_;
	return catdir( $self->splitted_dir, $self->authorbase_path_parts, $self->author_path_parts($author) );
}

sub modules {
	my ( $self ) = @_;
	my %modules;
	for (keys %{$self->packages->modules}) {
		$modules{$_} = catfile( $self->splitted_dir, $self->authorbase_path_parts, splitdir( $self->packages->modules->{$_}->[1] ) );
	}
	return \%modules;
}

sub timestamp { shift->packages->timestamp }

#
# Utilities
#

sub mkdir {
	my ( $self, @path ) = @_;
	make_path(catdir(@path),{ error => \my $err });
	if (@$err) {
		for my $diag (@$err) {
			my ($file, $message) = %$diag;
			if ($file eq '') {
				die "general error: $message\n";
			} else {
				die "problem making path $file: $message\n";
			}
		}
	}
}

1;

=encoding utf8

=head1 SYNOPSIS

  use CPAN::Repository;

  my $repo = CPAN::Repository->new({
    dir => '/var/www/greypan.company.org/htdocs/',
    url => 'http://greypan.company.org/',
  });
  
  $repo->initialize unless $repo->is_initialized;
  
  $repo->add_author_distribution('AUTHOR','My-Distribution-0.001.tar.gz');
  $repo->add_author_distribution('AUTHOR2','Other-Dist-0.001.tar.gz','Custom/Own/Path');
  $repo->set_alias('AUTHOR','The author <author@company.org>');
  $repo->set_alias('AUTHOR2','The other author <author@company.org>');
  
  my %modules = %{$repo->modules};
  
  my $fullpath_to_authordir = $repo->authordir('SOMEONE');
 
  my $packages = $repo->packages; # gives back a CPAN::Repository::Packages
  my $mailrc = $repo->mailrc; # gives back a CPAN::Repository::Mailrc

=head1 DESCRIPTION

This module is made for representing a directory which can be used as own CPAN for modules, so it can be a GreyPAN, a DarkPAN or even can be
used to manage a mirror of real CPAN your own way. Some code parts are taken from CPAN::Dark of B<CHROMATIC> and L<CPAN::Mini::Inject> of B<MITHALDU>.

=head1 SEE ALSO

L<CPAN::Repository::Packages>

L<CPAN::Repository::Mailrc>

=head1 SUPPORT

IRC

  Join #duckduckgo on irc.freenode.net. Highlight Getty for fast reaction :).

Repository

  http://github.com/Getty/p5-cpan-repository
  Pull request and additional contributors are welcome
 
Issue Tracker

  http://github.com/Getty/p5-cpan-repository/issues
